pkg/psalabelsyncer/sccrolecache.go

```diff
@@ -0,0 +1,550 @@
+	c.externalQueueEnqueueFunc = enqueueFunc
+	return c
+}
+
+func (c *SAToSCCCache) addToExternalJobQueue(obj interface{}) {
+	if c.externalQueueEnqueueFunc != nil {
+		c.externalQueueEnqueueFunc(obj)
+	}
+}
+
+func (c *SAToSCCCache) handleRoleModified(obj interface{}) {
+	sccs, err := c.sccLister.List(labels.Everything())
+	if err != nil {
+		klog.Warning("failed to list SCCs: %v", err)
+		return
+	}
+
+	role, err := NewRoleObj(obj)
+	if err != nil {
+		klog.Warningf("unexpected error, this may be a bug: %v", err)
+		return
+	}
+
+	c.usefulRolesLock.Lock()
+	defer c.usefulRolesLock.Unlock()
+	currentRoleSCCs := sets.NewString(c.usefulRoles[usefulRolesKey(role)]...)
+	c.syncRoleCache(role, role.Rules(), sccs)
+	updatedRoleSCCs := sets.NewString(c.usefulRoles[usefulRolesKey(role)]...)
+
+	// must be order-independent comparisons
+	if !currentRoleSCCs.Equal(updatedRoleSCCs) {
+		c.addToExternalJobQueue(role)
+	}
+}
+
+func (c *SAToSCCCache) handleRoleRemoved(obj interface{}) {
+	role, err := NewRoleObj(obj)
+	if err != nil {
+		klog.Warningf("unexpected error, this may be a bug: %v", err)
+		return
+	}
+
+	c.usefulRolesLock.Lock()
+	defer c.usefulRolesLock.Unlock()
+
+	if _, roleExists := c.usefulRoles[usefulRolesKey(role)]; roleExists {
+		delete(c.usefulRoles, usefulRolesKey(role))
+
+		c.addToExternalJobQueue(obj)
+	}
+}
+
+func (c *SAToSCCCache) handleSCCAdded(obj interface{}) {
+	scc, ok := obj.(*securityv1.SecurityContextConstraints)
+	if !ok {
+		klog.Errorf("expected object to be of type SecurityContextConstraints but it is %T", obj)
+		return
+	}
+
+	roles, err := c.roleLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("failed to handle %q SCC addition: %v", scc.Name, err)
+		return
+	}
+
+	clusterRoles, err := c.clusterRoleLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("failed to handle %q SCC addition: %v", scc.Name, err)
+		return
+	}
+
+	c.usefulRolesLock.Lock()
+	defer c.usefulRolesLock.Unlock()
+
+	for _, r := range roles {
+		role, err := NewRoleObj(r)
+		if err != nil {
+			panic(err)
+		}
+		c.syncRoleCacheSingleSCCAdded(role, r.Rules, scc)
+	}
+
+	for _, r := range clusterRoles {
+		role, err := NewRoleObj(r)
+		if err != nil {
+			panic(err)
+		}
+		c.syncRoleCacheSingleSCCAdded(role, r.Rules, scc)
+	}
+
+	c.addToExternalJobQueue(obj)
+}
+
+func (c *SAToSCCCache) handleSCCDeleted(obj interface{}) {
+	scc, ok := obj.(*securityv1.SecurityContextConstraints)
+	if !ok {
+		klog.Error("expected obj to be of type SecurityContextConstraints, got %T", obj)
+		return
+	}
+
+	c.usefulRolesLock.Lock()
+	defer c.usefulRolesLock.Unlock()
+
+	sccName := scc.Name
+	for role, sccs := range c.usefulRoles {
+		sccs := sccs
+
+		idxRemove := -1
+		for idx, cachedSCC := range sccs {
+			if sccName == cachedSCC {
+				idxRemove = idx
+				break
+			}
+		}
+		// remove the SCC from cache if it got removed
+		if idxRemove != -1 {
```

Comment:
complete stylistic nit: I believe conditions like this read better if expressed as `idxRemove < 0` but consider this just as a comment :-)

Reply:
the `idxRemove` is getting defined a few lines above with `-1` being the default value and so this condition is only checking whether it changed so I'd say this exact non-equals is a bit more accurate